Add SimpleConcurrentProcessor, clarify existing export concurrency (#4163)
open-telemetry/opentelemetry-specification: specification/logs/sdk.md

* `processor` - processor to be isolated.

> #### Simple Concurrent processor

Reviewer: This resonates with me too -- the use of "concurrent" suggests some kind of attention to concurrency, a limit of some sort in addition to offering concurrency. Maybe the name could be AsynchronousProcessor -- starts the export and returns success immediately.

I admit this worries me a bit -- is the exporter expected to implement some kind of limit, to avoid running out of memory? Can this be composed with the BatchSpanProcessor? What will control whether the BatchSpanProcessor drops spans vs. this processor consuming unlimited amounts of data?

Here, we have an OTel Collector processor that offers unlimited concurrency, but subject to a limit on the total amount of pending data, I consider it a potential solution to the problems posed above: https://github.com/open-telemetry/otel-arrow/tree/main/collector/processor/concurrentbatchprocessor

> #### Simple Concurrent processor

Author: > Maybe the name could be AsynchronousProcessor -- starts the export and returns success immediately.

That is *not* the intend. The intended exporters here not just 'starts' the export. It does everything inline (i.e serialization, writing to destination.), and then return success/failure. (from what I know, such exporters are typically writing to ETW, user-events)

> #### Simple Concurrent processor

Author: > is the exporter expected to implement some kind of limit, to avoid running out of memory?

No. Such exporters (etw/user-events) *does not* buffer anything. The logRecord is serialized and handed over to destination inline. ETW/user-events system is backed by OperatingSystem kernel memory, and they have ample mechanisms to keep memory in control, but those are outside the scope of the exporter.

> #### Simple Concurrent processor

Author: naming is hard 🤣 I am not sure if any of the alternate suggestions are significantly better. I'll keep thinking and hope we get more suggestions.

that will implement this function are typically expected to serialize and
transmit the data to the destination.

`Export` MAY be called concurrently for the same exporter instance if paired
with [Simple Concurrent Processor](#simple-concurrent-processor). Each exporter

> implementation MUST document whether it supports concurrent `Export` calls, and

Reviewer 2: 1. We cannot force custom implementation on how they document their implementations
2. It is not only about exporting but also ForceFlush and Shutdown

```suggestion
implementation provided by the SDK MUST document whether it is concurrent safe, and
```

Author: We cannot enforce, but the spec wording is to make sure any person authoring own exporter should follow this spec. If they don't follow, then there could be undesired behavior.
Alternatively, we can word it in such a way that "It must be documented to the exporter authors that they should document the exporters' concurrency characteristics....

Reviewer 3: I believe this was part of why it was went with that the processor should called all exporters the same and not have to worry about concurrency.

Reviewer 2: @tsloughter, the users could still use processors that does not require the exporters to be concurrent safe. But we would give a more performant option for cases where exporter are concurrent safe and can be used synchronously.

@cijothomas, the alternative could be that such exporter packages (like ETW or user_events exporters) would provide their own processors which are "tailored" to their exporters. It can give more flexibility and not increase the SDK functionality surface.

Author: > the alternative could be that such exporter packages (like ETW or user_events exporters) would provide their own processors which are "tailored" to their exporters

You are right! That is exactly what we have been doing for years. But that does not mean spec shouldn't support them.
The increase in SDK functionality surface is necessary, to support such scenarios. The spec already has wording that state implementations must have simple and batch (meaning others are optional), so additional processors don't put extra burden on implementations, unless they chose to support.

Reviewer 2: What is more (I have forgotten tot call it out explictly) I think that e.g. for ETW we can simply provide a single `EtwProcessor` which does the exporting. There is no need for batching or synchronization so there is no need to provide an `Exporter` interface implementation.

Author: > I think that e.g. for ETW we can simply provide a single EtwProcessor which does the exporting

Yes that is also possible. (In certain ways, OTel Rust does that. Its etwexporter is not really an exporter, but a ~processor).
But I think it is best to model exporters (the thing which does serialization, export telemetry to outside the process) as exporters itself consistently.

Reviewer 2: > But I think it is best to model exporters

I am not really sure as these exporters are emitting batches. And for use cases like ETW and UserEvents you probably prefer to operate on "single" log record.

Author: They are exporters! (They do the job of serializing and transferring the telemetry to an external entity.) Whether it gets a batch of 1 item or multiple is purely based on choice of processor used. (when used with SimpleProcessor, even OTLPExporter gets a batch of single item only).

hence can be paired with the [Simple Concurrent
Processor](#simple-concurrent-processor).

Depending on the implementation, the result of the export may be returned to the
Processor not in the return value of the call to `Export` but in a language
specific way for signaling completion of an asynchronous task. This means that
while an instance of an exporter that does not support concurrent calls will
never have `Export` called concurrently, it does not mean that the task of
exporting cannot be done concurrently. How this is done is outside the scope of

> this specification. Each implementation MUST document the concurrency
> characteristics the SDK requires of the exporter.

Reviewer 2: I think that after the proposed changes this sentence is unnecessary and could only bring confusion.

```suggestion
this specification.
```

that will implement this function are typically expected to serialize and
transmit the data to the destination.

`Export` MAY be called concurrently for the same exporter instance if paired
with [Simple Concurrent Processor](#simple-concurrent-processor). Each exporter
implementation MUST document whether it supports concurrent `Export` calls, and
hence can be paired with the [Simple Concurrent

> Processor](#simple-concurrent-processor).

Reviewer 4: Wondering whether documentation the Exporter alone is sufficient, or if we need to introduce a guard mechanism, such as having the Exporter expose an API like `can_handle_concurrent_exports()`. As using a non-concurrent exporter with a concurrent processor could be dangerous, leading to race conditions, data corruption, and other issues. Though this MAY be a breaking change for existing exporter implementations.

Reviewer 2: > Though this MAY be a breaking change for existing exporter implementations.

It would for OTel Go SDK.

Author: @lalitb Nothing is preventing language implementations to enforce that mechanism. I am trying to keep spec simpler.

One example .NET is trying to follow: The exporter never exposes its constructor, so nobody can instantiate one. Exporter instead can offer static methods like "GetExporterWithCorrectProcessor", and the exporter can return an instance of itself wrapped inside the correct Processor it supports, ensuring users cannot accidentally pair wrong processor with exporter. 

I feel its too much implementation details to call out in the spec.

Depending on the implementation, the result of the export may be returned to the
Processor not in the return value of the call to `Export` but in a language
specific way for signaling completion of an asynchronous task. This means that
while an instance of an exporter that does not support concurrent calls will
never have `Export` called concurrently, it does not mean that the task of

> exporting cannot be done concurrently. How this is done is outside the scope of
> this specification. How this is done is outside the scope of this specification.
> Each implementation MUST document the concurrency characteristics the SDK
> requires of the exporter.

Reviewer 2: Repeated sentence

```suggestion
exporting cannot be done concurrently. How this is done is outside the scope of
this specification. Each implementation MUST document the concurrency
characteristics the SDK requires of the exporter.
```
